presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java

```diff
@@ -166,4 +222,47 @@ private static void skipRequestBody(HttpServletRequest request)
             copy(inputStream, nullOutputStream());
         }
     }
+
+    public static class ModifiedHttpServletRequest
+            extends HttpServletRequestWrapper
+    {
+        private final Map<String, String> customHeaders;
+
+        public ModifiedHttpServletRequest(HttpServletRequest request, Map<String, String> headers)
+        {
+            super(request);
+            this.customHeaders = new ConcurrentHashMap<>(headers);
+        }
+
+        @Override
+        public String getHeader(String name)
+        {
+            String headerValue = customHeaders.get(name);
+            if (headerValue != null) {
+                return headerValue;
+            }
+            return super.getHeader(name);
+        }
+
+        @Override
+        public Enumeration<String> getHeaderNames()
+        {
+            ImmutableSet.Builder<String> headerNamesBuilder = ImmutableSet.builder();
+            headerNamesBuilder.addAll(customHeaders.keySet());
+            Enumeration<String> originalHeaderNames = super.getHeaderNames();
+            while (originalHeaderNames.hasMoreElements()) {
+                headerNamesBuilder.add(originalHeaderNames.nextElement());
+            }
+            return Collections.enumeration(headerNamesBuilder.build());
+        }
+
+        @Override
+        public Enumeration<String> getHeaders(String name)
+        {
+            if (customHeaders.containsKey(name)) {
+                return Collections.enumeration(Collections.singleton(customHeaders.get(name)));
```

Comment:
```suggestion
                return enumeration(immutableListOf(customHeaders.get(name)));
```

Please static import as described in our contributing guide.